### ISSUE TITLE
About the plot 3a in the paper "Intra-modal and inter-modal CLIP cosine similarities"

### ISSUE THREAD

**Issue by XuRui314** [vishaal27/SuS-X#3]
I wonder how this figure is ploted, take Image-Text for example, do you calculate every combination of image and text similarities? And is there any notebook demo about this, thx.

**Comment by vishaal27** [comment]
Hi, thanks for the question.
For Fig. 3a, we take ImageNet's test set. For every pair of image-text samples, we compute the pairwise image-image, text-text, and image-text cosine similarities. There are 50000 image samples (since that's the size of the test-set) and there are 1000 text samples (since that's the number of classes). To be more concrete, you can use this code for generating the plot:

```python
def cos_sim_vector_with_others(vector, other_vectors, eps=1e-8):
    vector_n = vector.norm(dim=1)[:, None]
    other_vectors_n = other_vectors.norm(dim=1)[:, None]
    vector = vector / torch.max(vector_n, eps * torch.ones_like(vector_n))
    other_vectors = other_vectors / torch.max(other_vectors_n, eps * torch.ones_like(other_vectors_n))
    cos_sim_matrix = torch.mm(vector, other_vectors.transpose(0, 1))
    return cos_sim_matrix

# This assumes that:
# zeroshot_weights are the zero-shot text classifier weights of imagenet, dim (c,d), where c is 1000 for imagenet and d is 1024 for CLIP-RN50
# test_features are the image features over imagenet test-set, dim (n,d) where n is 50000 for imagenet
 
intra_text_sims = cos_sim_vector_with_others(zeroshot_weights, zeroshot_weights)
intra_image_sims = cos_sim_vector_with_others(test_features, test_features)
inter_image_text_sims = cos_sim_vector_with_others(zeroshot_weights, test_features)
# mask out diagional elements to ensure sims are not highly skewed due to 1-similarities
intra_text_sims.fill_diagonal_(0)
intra_image_sims.fill_diagonal_(0)

true_diag_sims = []
for id, ten in enumerate(inter_image_text_sims):
true_diag_sims = true_diag_sims + list(ten[id*50 : (id+1)*50].cpu().numpy())
true_diag_sims = np.asarray(true_diag_sims)

import matplotlib.pyplot as plt

plt.hist(intra_text_sims.flatten().cpu().numpy(), bins=20, label='Text-only similarities', density=True, alpha=0.6)
plt.hist(intra_image_sims.flatten().cpu().numpy(), bins=20, label='Image-only similarities', density=True, alpha=0.6)
plt.hist(inter_image_text_sims.flatten().cpu().numpy(), bins=20, label='Image-text similarities', density=True, alpha=0.6)    
plt.hist(torch.tensor(true_diag_sims).flatten().cpu().numpy(), bins=20, label='Paired Image-text similarities', density=True, alpha=0.6)    

plt.xlabel("Cosine Similarities")
plt.ylabel("Sample Density")
plt.legend(loc='best')
plt.tight_layout()
plt.show()

### Validation to check you're using the right features: this should give ~60.31% accuracy
logits = 100. * test_features @ zeroshot_weights.T
labels = test_labels
np_preds = torch.argmax(logits, dim=1).cpu().numpy()
np_labels = labels.cpu().numpy() 
acc = 100*(np_preds == np_labels).sum()/np_labels.shape[0]
print('ZS Acc for Imagenet = ', ensemble_acc)
```

I could write this up in a demo colab notebook if I get the time soon. However, please let me know if you have any questions with this code, or are unable to reproduce the plot, happy to help further :)

**Comment by XuRui314** [comment]
Really thanks for sharing :)
